### ISSUE TITLE
Support "antialiased" splats

### ISSUE THREAD

**Issue by bchretien** [mkkellogg/GaussianSplats3D#142]
Hi!

First of all, thanks for the project and kudos for the code quality and the comments (e.g. in `SplatMesh.js`). I was checking your viewer after playing around with Nerfstudio. I enabled the new "antialiasing" version of their rasterizer (see motivation [here](https://github.com/nerfstudio-project/gsplat/pull/117) and integration in Nerfstudio [there](https://github.com/nerfstudio-project/nerfstudio/pull/2888)), but the problem is that it requires explicit support in viewers down the line. To summarize, a compensation factor would need to be computed in the vertex shader and applied to the gaussian opacities.

Would that feature be in the scope of this project? If it is, I could provide a PR. Since rasterization is done differently in this viewer, I guess the compensation would need to be applied to `vColor.a` directly?

**Comment by mkkellogg** [comment]
This seems like it would be a good feature to have, but I'm not sure how best to integrate it. Specifically I'd like to avoid having to specify a parameter when loading a file to indicate whether it is an "anti-aliased" or "classic" splat. Ideally that information would be encoded into the file itself -- Does the Nerfstudio code support  that? 

As for updating my viewer, it looks like it would be super easy, in fact the authors of the [mip-splatting](https://github.com/autonomousvision/mip-splatting/tree/746a17c9a906be256ed85b8fe18632f5d53e832d) work created a web viewer (based on mine) for that purpose, and it looks like they just made a small update to the vertex shader in `SplatMesh.js`:

```
  // compute the coef of alpha based on the detemintant
  float kernel_size = 0.1;
  float det_0 = max(1e-6, cov2Dm[0][0] * cov2Dm[1][1] - cov2Dm[0][1] * cov2Dm[0][1]);
  float det_1 = max(1e-6, (cov2Dm[0][0] + kernel_size) * (cov2Dm[1][1] + kernel_size) - cov2Dm[0][1] * cov2Dm[0][1]);
  float coef = sqrt(det_0 / (det_1+1e-6) + 1e-6);

  if (det_0 <= 1e-6 || det_1 <= 1e-6){
      coef = 0.0f;
  }

  cov2Dm[0][0] += kernel_size;
  cov2Dm[1][1] += kernel_size;
  vColor.a *= coef;
```

**Comment by ichsan2895** [comment]
Hello, I want to fork this repo and add this code, but I don't know where to put the code:
```
  // compute the coef of alpha based on the detemintant
  float kernel_size = 0.1;
  float det_0 = max(1e-6, cov2Dm[0][0] * cov2Dm[1][1] - cov2Dm[0][1] * cov2Dm[0][1]);
  float det_1 = max(1e-6, (cov2Dm[0][0] + kernel_size) * (cov2Dm[1][1] + kernel_size) - cov2Dm[0][1] * cov2Dm[0][1]);
  float coef = sqrt(det_0 / (det_1+1e-6) + 1e-6);

  if (det_0 <= 1e-6 || det_1 <= 1e-6){
      coef = 0.0f;
  }

  cov2Dm[0][0] += kernel_size;
  cov2Dm[1][1] += kernel_size;
  vColor.a *= coef;
```

Anyway, If we apply antialiased method for all PLY. May it broken any PLY that created from other software? (Inria 3DGS, Luma, Polycam, OpenSplat, etc) ?

**Comment by jb-ye** [comment]
I am the author of antialiasing mode of splatfacto , the code change needed to support web viewer is small (but slightly different from mipsplatting). 

https://github.com/nerfstudio-project/gsplat/blob/main/gsplat/_torch_impl.py#L188

Currently it is not possible to store any metadata in PLY format. Any suggestions for us to move forward? 

For short term, it makes sense to have a keyboard hotkey to toggle between classic and antialiasing mode.

**Comment by mkkellogg** [comment]
@jb-ye It would definitely be helpful if you could provide a demo file. Then I can take a stab at implementing the computation for the opacity compensation factor.

**Comment by jb-ye** [comment]
Will do so when I got time this week. @bchretien @mkkellogg

**Comment by jb-ye** [comment]
https://drive.google.com/file/d/19e0iAsoc9F26ilM4s6g0Y6n0n9-WRjV5/view?usp=drive_link
Please check out this link for a sample ply asset.

Left: classic mode rendering of antialiased asset; Right: antialiased mode
<img width="821" alt="Screenshot 2024-03-21 at 2 05 08 PM" src="https://github.com/mkkellogg/GaussianSplats3D/assets/132313008/01987a54-2afa-4768-8ce6-81a47ed32115">

**Comment by mkkellogg** [comment]
@jb-ye Thanks! I'll try to make the update in the next couple of days and let you know when it's ready.

**Comment by mkkellogg** [comment]
@jb-ye Would you be able to share the `cameras.json` (or equivalent parameters) for the above scene?

**Comment by jb-ye** [comment]
@mkkellogg What do you mean by `cameras.json`. Are they training camera parameters?

**Comment by mkkellogg** [comment]
Well I noticed that using the standard calculation for focal length using Three.js projection matrix is producing sub-optimal results in terms of render quality (subject for sure), so I thought I'd play around with them to see if I can get better results by matching the training camera parameters.

**Comment by jb-ye** [comment]
Just want to double check your formula implemented. Here is the reference implementation from splatfacto/gsplat lib, where blurring kernel is 0.3 pixel and compensation is clamped instead of adding a tiny round-off eps. This is different from the one used by mip-splatting. 
```
    det_orig = cov2d[..., 0, 0] * cov2d[..., 1, 1] - cov2d[..., 0, 1] * cov2d[..., 0, 1]
    cov2d[..., 0, 0] = cov2d[..., 0, 0] + 0.3
    cov2d[..., 1, 1] = cov2d[..., 1, 1] + 0.3
    det_blur = cov2d[..., 0, 0] * cov2d[..., 1, 1] - cov2d[..., 0, 1] * cov2d[..., 0, 1]
    compensation = torch.sqrt(torch.clamp(det_orig / det_blur, min=0))
```

**Comment by mkkellogg** [comment]
Yep that's that math I'm using. I got something working that I think looks pretty good:

![before_after](https://github.com/mkkellogg/GaussianSplats3D/assets/6249005/3220d044-f98e-426b-b1ce-ed9a19188d92)


I used this code in the shader: 
```
  float detOrig = cov2Dm[0][0] * cov2Dm[1][1] - cov2Dm[0][1] * cov2Dm[0][1];
  cov2Dm[0][0] += 0.3;
  cov2Dm[1][1] += 0.3;
  float detBlur = cov2Dm[0][0] * cov2Dm[1][1] - cov2Dm[0][1] * cov2Dm[0][1];
  float compensation = sqrt(max(detOrig / detBlur, 0.0));
  vColor.a *= compensation;
```

**Comment by jb-ye** [comment]
Fantastic

**Comment by mkkellogg** [comment]
Thank you for your help on this!

**Comment by gonzalle** [comment]
That is a great news ! Can't wait to test it !

**Comment by ichsan2895** [comment]
My experiment with `Truck` dataset

Left = Splatfacto-big with classic mode
Right = Splatfacto-big with antialiased mode

![A0000_GSplat018-Antialiased-SplatBIG](https://github.com/mkkellogg/GaussianSplats3D/assets/60387342/aba734cf-44b1-4b18-9714-79c3126483a6)

![A0001_GSplat018-Antialiased-SplatBIG](https://github.com/mkkellogg/GaussianSplats3D/assets/60387342/4e4742dc-27cd-4097-bc64-7af38d3817c9)

**Comment by jb-ye** [comment]
@mkkellogg thanks if you want to use half precision to store covariance, you would notice some quality loss for antialiasing mode. The fix is instead of saving the covariance in half precision, saving its Cholesky decomposition in half precision, and multiple it back to get the full covariance of full precision at usage time.

**Comment by gonzalle** [comment]
I have fantastic results with AA splats but still some issues with highly contrasted models
from a "standard" pov, (close to the training images), the details are undoubtly better.
Left : No AA , right with AA
![aa_close](https://github.com/mkkellogg/GaussianSplats3D/assets/1769103/e922330e-42ea-4aed-8c68-82fabbf6e308)

But when I get my camera further, I get some weird white artifacts on the AA model.
(the non AA model still looks bad but don't have those artifacts.
![aa_far](https://github.com/mkkellogg/GaussianSplats3D/assets/1769103/3ac502d2-ba0f-494f-8d76-d8afb71531eb)

Any idea about this problem ?... I guess it has something to do with transparency, since whats is behind the fabric in my model is white.

**Comment by jb-ye** [comment]
@gonzalle did you observe similar issue in nerfstudio viewer?

**Comment by gonzalle** [comment]
> @gonzalle did you observe similar issue in nerfstudio viewer? I think it might relate to spherical harmonics not used in the web viewer.

Indeed ! 
the nerfstudio viewer rendering was clean from any distance...
Is there any trick around ? beside having less textured material ?

**Comment by jb-ye** [comment]
> > @gonzalle did you observe similar issue in nerfstudio viewer? I think it might relate to spherical harmonics not used in the web viewer.
> 
> Indeed ! the nerfstudio viewer rendering was clean from any distance... Is there any trick around ? beside having less textured material ?

Maybe try to reduce directional lights during capture, use more ambient lighting

**Comment by mkkellogg** [comment]
@gonzalle Would you be willing to share your model? Maybe I can do some troubleshooting on my end.

**Comment by gonzalle** [comment]
> @gonzalle Would you be willing to share your model? Maybe I can do some troubleshooting on my end.

Certainly, I just have to train again my model...
I hope it'll be fast... gime a couple hours :))

**Comment by gonzalle** [comment]
[aa_weird.zip](https://github.com/mkkellogg/GaussianSplats3D/files/14796467/aa_weird.zip)
Mark,
Here is the file... As I had to re-train it and edit it, I noticed that nerfstudio (when at low res) and Supersplat, both show the same issue.
For me it's something that has to do with the big white splat behind the fabric texture and some kind of "mip mapping" issues...

**Comment by mkkellogg** [comment]
So I have also noticed this issue, and I don't know exactly what is happening but it does seem like it's caused by splats that are much bigger than the other splats in the scene. I added a super hacky viewer parameter called `focalAdjustment` that I have discovered helps. Try something like this when you run the viewer: 

```
const viewer = new GaussianSplats3D.Viewer({
  'focalAdjustment': 3.0
});
```

The default value for `focalAdjustment` is 1.0, increasing it seems to make these kinds of artifacts less noticeable. I want to reiterate this is a big ol' hack, and I don't know exactly why it helps :)

**Comment by gonzalle** [comment]
Great, I will try that. Thanks a lot !
Gaussian Splatting, as a global techniqu,e and as it stands today, still has some issues. In any case, I firmly believe that such a technique, combined with a good web viewer, is a fantastic step forward! Keep on the fantastic job !
